### ISSUE TITLE
getpwnam(): name not found: foo when trying to login

### ISSUE THREAD

**Issue by GSMCD** [jupyterhub/ldapauthenticator#28]
Hello folks,

When trying to login using ldapauthenticator I'm running into the following issue:

I attempt to login to Jupyterhub with LDAP user _foo_, it returns: 

> 500 : Internal Server Error
Failed to start your server. Please contact admin.

**Log Output**
```
    Traceback (most recent call last):
      File "/usr/local/anaconda3/lib/python3.5/site-packages/tornado/web.py", line 1469, in _execute
        result = yield result
      File "/usr/local/anaconda3/lib/python3.5/site-packages/jupyterhub/handlers/login.py", line 79, in post
        yield self.spawn_single_user(user)
      File "/usr/local/anaconda3/lib/python3.5/site-packages/jupyterhub/handlers/base.py", line 312, in spawn_single_user
        yield gen.with_timeout(timedelta(seconds=self.slow_spawn_timeout), f)
      File "/usr/local/anaconda3/lib/python3.5/site-packages/jupyterhub/user.py", line 247, in spawn
        raise e
      File "/usr/local/anaconda3/lib/python3.5/site-packages/jupyterhub/user.py", line 228, in spawn
        yield gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
      File "/usr/local/anaconda3/lib/python3.5/types.py", line 243, in wrapped
        coro = func(*args, **kwargs)
      File "/usr/local/anaconda3/lib/python3.5/site-packages/jupyterhub/spawner.py", line 471, in start
        preexec_fn=self.make_preexec_fn(self.user.name),
      File "/usr/local/anaconda3/lib/python3.5/site-packages/jupyterhub/spawner.py", line 417, in make_preexec_fn
        return set_user_setuid(name)
      File "/usr/local/anaconda3/lib/python3.5/site-packages/jupyterhub/spawner.py", line 374, in set_user_setuid
        user = pwd.getpwnam(username)
    KeyError: 'getpwnam(): name not found: foo'
```

I'm guessing this is down to getpwnam() looking for the user _foo_ in the /etc/passwd file. However as foo is an LDAP user it does not exist there. 

**jupyterhub_config.py**

```
c.JupyterHub.authenticator_class = 'ldapauthenticator.LDAPAuthenticator'
c.LDAPAuthenticator.server_address = 'ldap-server'
c.LDAPAuthenticator.bind_dn_template = 'cn={username},ou=Team1,l=Europe,O=thing'
```

I'm at a loss as to how to resolve this, any help would be greatly appreciated.

**Comment by julien-marchand** [comment]
Hi guys,

Any update ?

I'm stuck with the same error, 
`Message: "Unhandled error starting foo bar's server: 'getpwnam(): name not found: foo bar'"`

With the following settings :
`c.JupyterHub.authenticator_class = 'ldapauthenticator.LDAPAuthenticator'
c.LDAPAuthenticator.server_address = 'internal.domain.name'
c.LDAPAuthenticator.bind_dn_template = '{username}' # Same error with full DN
c.LDAPAuthenticator.valid_username_regex = '^[a-zA-Z][\s.a-zA-Z0-9_-]*$'
c.LDAPAuthenticator.use_ssl = False
c.LDAPAuthenticator.lookup_dn = True
c.LDAPAuthenticator.user_search_base = 'OU=foo,DC=bar'
c.LDAPAuthenticator.user_attribute = 'sAMAccountName'`

Thank you in advance :)

**Comment by GSMCD** [comment]
Hi Julien,

I actually resolved this, it was a strange issue with jupyterhub spawner.py.

We would be trying to login as "A12345" and spawner.py would throw an error saying:

`KeyError: 'getpwnam(): name not found: a12345'`

Note the lowercase "a", our LDAP server was expecting a capital "A". 

What we ended up doing to get it working was add a few lines in spawner.py to uppercase the username when it detected the first character being "a"

It wasn't ideal, but it got it working.
